packages/core/src/domain/conversion/upload-conversion-steps.ts

```diff
@@ -0,0 +1,220 @@
+import { Bundle, Resource } from "@medplum/fhirtypes";
+import { errorToString } from "@metriport/shared";
+import { SQSRecord } from "aws-lambda";
+import { S3Utils, executeWithRetriesS3 } from "../../external/aws/s3";
+import { capture } from "../../util";
+import { FHIR_APP_MIME_TYPE, XML_APP_MIME_TYPE } from "../../util/mime";
+import { buildDocumentNameForPartialConversions } from "./filename";
+
+export const defaultS3RetriesConfig = {
+  maxAttempts: 3,
+  initialDelay: 500,
+};
+
+export async function storePreProcessedConversionResult({
+  s3Utils,
+  conversionResult,
+  conversionResultBucketName,
+  conversionResultFilename,
+  message,
+  context,
+  lambdaParams,
+  log,
+}: {
+  s3Utils: S3Utils;
+  conversionResult: Bundle<Resource>;
+  conversionResultBucketName: string;
+  conversionResultFilename: string;
+  message: SQSRecord;
+  context: string;
+  lambdaParams: Record<string, string | undefined>;
+  log: typeof console.log;
+}) {
+  try {
+    await storeInS3WithRetries({
+      s3Utils,
+      payload: JSON.stringify(conversionResult),
+      bucketName: conversionResultBucketName,
+      fileName: conversionResultFilename,
+      contentType: FHIR_APP_MIME_TYPE,
+      log,
+    });
+  } catch (error) {
+    const msg = "Error uploading converted FHIR Bundle";
+    log(`${msg}: ${error}`);
+    capture.error(msg, {
+      extra: {
+        message,
+        ...lambdaParams,
+        conversionResultFilename,
+        context,
+        error,
+      },
+    });
+  }
+}
+
+export async function storePartitionedPayloadsInS3({
+  s3Utils,
+  partitionedPayloads,
+  conversionResultBucketName,
+  preConversionFilename,
+  message,
+  context,
+  lambdaParams,
+  log,
+}: {
+  s3Utils: S3Utils;
+  partitionedPayloads: string[];
+  conversionResultBucketName: string;
+  preConversionFilename: string;
+  message: SQSRecord;
+  context: string;
+  lambdaParams: Record<string, string | undefined>;
+  log: typeof console.log;
+}) {
+  partitionedPayloads.forEach(async (payload, index) => {
```

Comment:
I see we moved this here from another file, but this is problematic: all code executed in a lambda must be completed before the lambda finishes, and this will return the execution to the lambda before it actually gets completed, since each iteration of the `.forEach()` runs a Promise.

Reply:
It should prob be `await executeAsynchronously(...)`.